### PR TITLE
Botany stuff

### PR DESCRIPTION
**Pull request by ThunderBear2006**
## About the PR

Added a new system where plants may transmutate into whole new plant species.

The way this happens is through a new plant gene: TRA(nsmution), TRA is three integer values
that can mutate like other plant genetics (e.g. plant yield).

When a plant's TRA value matches one of its transmutations TRA values it transmutes into that plant,
Plant transmutations are defined in their seed prototype, meaning a lemon plant won't turn into an apple tree (unless you add that transmutation).

Added new plants that can only be acquired through plant transmutation:

Cannabis can turn into: 
Rainbow cannabis, grinds into THC, Space drugs and mind breaker toxin

Banana tree can turn into:
Pink banana tree: produces pink bananas that can be ground or juiced into a new chemical: Laughtex (more on that later)

Lemon plant can turn into:
Hell Lemon: Has a very short fuse that triggers a tiny but powerful explosion, can either go off by being used or at random

Poppy can turn into:
Clown flower: works like a spray bottle but with a higher velocity, only holds 10 units

Fly Amanita can turn into:
Shroominal Pod: Produces a Shroomkin (mushroom chicken) that lays Shreggs (shroom eggs), Shreggs are like normal eggs but contain a significant amount of blood and some space drugs in them

Shroominal Pod can turn into:
Shroomanoid Pod: Produces a Shrooman (mushroom man), a new ghost role, Shroomen can only wear a belt, a backpack, gloves, neck garments, mask, glasses and ear pieces. Shroomen talk with an elegant accent (like if you read a thesaurus as a pass time).

Shroomen are very weak to fire, weak to slash, blunt and shock, but take very little asphyxiation damage and cold damage and take half poison damage. Bare handed attacks deal 2 brute and 6 poison, so great if the station is spaced but bad in a fight

Laughtex: A new toxin that makes the subject laugh uncontrollably (I swear I've typed this sentence so many damn times) but slowly suffocates them, deals 3 units of airloss scaled by quantity, same effects if touched but only 1 unit of airloss scaled by quantity.

TRA-05: Chemical that increments plant TRA values in the positive by one unit when metabolized at the cost of using 10 units of water and nutriment

TRA-91: Like TRA-05 but decrements TRA, costs the plant half the nutriment and water

**Media**

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

All plants and their produce:
image

Laughtex Effects:
https://user-images.githubusercontent.com/100388962/234360766-557e2f09-eca9-4b63-93ee-1fa9fb1855d5.mp4

Fully decked out shrooman:
image
image
image

Shrooman inventory slots:
image

Hell Lemon in use:
https://user-images.githubusercontent.com/100388962/234363426-af82de19-45b6-4659-9010-560a7d4cfcad.mp4

**Changelog**

:cl: ThunderBear#9834
- add: The egg heads over at Centcom have discovered a new element to plant genetics, they're calling it the TRA Sequence and plants with the correct TRA Sequence have been documented to transmute into new plant species! Centcom has advised all botanists and scientists to experiment with this new genetic element in hopes of discovering new plant species!
- add: Added Laughtex, a chemical reagent that causes the subject to laugh uncontrollably during metabolization or touch, causes airloss damage and comes from pink bananas!
- add: Added TRA-05 and TRA-91, two new chemical reagents that can increment or decrement (respectively) a plant's TRA sequence during metabolization

### CODE REVIEW

**Comment by ThunderBear2006** [comment]
Tiny note: If anyone reviews this requesting a change please @me on the discord server

**Comment by potato1234x** [comment]
missed opportunity to call the hell lemons combustible lemons

**Comment by metalgearsloth** [comment]
> Shroominal Pod
What's wrong with pod people?

Additionally I'm hesitant to add even more botany values when the seedprototype needs a refactor and none of it is exposed to players still.

**Comment by ThunderBear2006** [comment]
> missed opportunity to call the hell lemons combustible lemons

I thought it would line up with the quote better

**Comment by ThunderBear2006** [comment]
> > Shroominal Pod
> > What's wrong with pod people?
> 
> Additionally I'm hesitant to add even more botany values when the seedprototype needs a refactor and none of it is exposed to players still.

That is a shame, I'm not sure what I could do for the seed prototype but I maybe I could merge my PR with the plant analyzer PR if thats possible?

**Comment by ThunderBear2006** [comment]
> > > Shroominal Pod
> > > What's wrong with pod people?
> > 
> > 
> > Additionally I'm hesitant to add even more botany values when the seedprototype needs a refactor and none of it is exposed to players still.
> 
> That is a shame, I'm not sure what I could do for the seed prototype but I maybe I could merge my PR with the plant analyzer PR if thats possible?

Also I didn't make pod people because we already have plant people, the Diona. I thought maybe a different approach could be taken while achieving a similar result

File: Content.Server/Botany/Systems/BotanySystem.Seed.cs [hunk 1]
```diff
@@ -86,28 +86,34 @@ private void OnExamined(EntityUid uid, SeedComponent component, ExaminedEvent ar
...
+        args.PushMarkup("A: " + seed.TRA.A);
     }
 
     #region SeedPrototype prototype stuff
 
     /// <summary>
     /// Spawns a new seed packet on the floor at a position, then tries to put it in the user's hands if possible.
     /// </summary>
-    public EntityUid SpawnSeedPacket(SeedData proto, EntityCoordinates coords, EntityUid user)
+    public EntityUid SpawnSeedPacket(SeedData seedData, EntityCoordinates coords, EntityUid user, bool sampled = false)
```
Comment (vulppine) [inline Content.Server/Botany/Systems/BotanySystem.Seed.cs:101]: unused parameter

File: Content.Server/Chemistry/ReagentEffects/PlantMetabolism/PlantShiftTRA.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+using Content.Server.Botany.Systems;
+using Content.Shared.Chemistry.Reagent;
+using JetBrains.Annotations;
+
+namespace Content.Server.Chemistry.ReagentEffects.PlantMetabolism
```
Comment (vulppine) [inline Content.Server/Chemistry/ReagentEffects/PlantMetabolism/PlantShiftTRA.cs:5]: this needs to be a filescoped namespace

File: Content.Server/Explosion/Components/TriggerRandomlyComponent.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+namespace Content.Server.Explosion.Components
```
Comment (vulppine) [inline Content.Server/Explosion/Components/TriggerRandomlyComponent.cs:1]: this needs to be a filescoped namespace

File: Content.Server/Explosion/EntitySystems/TriggerSystem.cs [hunk 4]
```diff
@@ -238,5 +241,27 @@ private void UpdateTimer(float frameTime)
                     _appearance.SetData(uid, TriggerVisuals.VisualState, TriggerVisualState.Unprimed, appearance);
             }
         }
+
+        private void UpdateRandom(float frameTime){
```
Comment (vulppine) [inline Content.Server/Explosion/EntitySystems/TriggerSystem.cs:245]: fix the braces here

File: Content.Server/Explosion/EntitySystems/TriggerSystem.cs [hunk 5]
```diff
@@ -238,5 +241,27 @@ private void UpdateTimer(float frameTime)
...
+            foreach (var random in EntityQuery<TriggerRandomlyComponent>()){
```
Comment (vulppine) [inline Content.Server/Explosion/EntitySystems/TriggerSystem.cs:246]: fix the braces here

File: Content.Shared/Botany/TransmutationPrototype.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+using Robust.Shared.Prototypes;
+
+namespace Content.Server.Botany;
+
+[Prototype("transmutation")]
+public sealed class TransmuationPrototype : IPrototype
+{
+    [IdDataField] public string ID { get; private init; } = default!;
+
+    [DataField("plantPrototype")] public string prototype = "";
```
Comment (vulppine) [inline Content.Shared/Botany/TransmutationPrototype.cs:10]:
```suggestion
    [DataField("plantPrototype")] public string Prototype = "";
```

File: Content.Server/Botany/Systems/PlantHolderSystem.cs [hunk 7]
```diff
@@ -756,6 +768,29 @@ public void RemovePlant(EntityUid uid, PlantHolderComponent? component = null)
             UpdateSprite(uid, component);
         }
 
+        /// <summary>
+        ///     Checks if the provided seed data's TRA matches any of it's transmutation prototypes' TRA values
+        /// </summary>
+        public bool TryToTransmutate(SeedData seedData, [NotNullWhen(true)] out SeedData? seed){
+            foreach(var tra_proto in seedData.PlantTransmutations){
```
Comment (vulppine) [inline Content.Server/Botany/Systems/PlantHolderSystem.cs:775]: variable needs to be `camelCase`, so:

```suggestion
            foreach (var traProto in seedData.PlantTransmutations)
            {
```

**Review by vulppine — CHANGES_REQUESTED** [review]
there's a lot of formatting issues, please make sure that your bracing adheres to this style:

```cs
blockStatement
{
  // lines go here
}
```

i've attached a few review comments to some code, but generally you'll need to go over your changed files and fix this up

**Comment by vulppine** [comment]
> > > > Shroominal Pod
> > > > What's wrong with pod people?
> > > 
> > > 
> > > Additionally I'm hesitant to add even more botany values when the seedprototype needs a refactor and none of it is exposed to players still.
> > 
> > 
> > That is a shame, I'm not sure what I could do for the seed prototype but I maybe I could merge my PR with the plant analyzer PR if thats possible?
> 
> Also I didn't make pod people because we already have plant people, the Diona. I thought maybe a different approach could be taken while achieving a similar result

diona in SS13 *are* the pod people (at least in some implementations of SS13), so i think that you should consider making it so that the pods spawn them instead

**Comment by ThunderBear2006** [comment]
> diona in SS13 _are_ the pod people (at least in some implementations of SS13), so i think that you should consider making it so that the pods spawn them instead

I would have done that but I thought it would've been better if the humanoid that was spawned wouldn't be as good as a round start species, otherwise you'd just be able to respawn once the botanist whips up a batch of them, at least with the current implementation there are downsides.

Either way with @metalgearsloth and others not wanting to add more stuff to botany until it is refactored makes me think its not worth the time to make changes to this PR unless there is a solution that makes everyone happy

**Comment by vulppine** [comment]
> > diona in SS13 _are_ the pod people (at least in some implementations of SS13), so i think that you should consider making it so that the pods spawn them instead
> 
> I would have done that but I thought it would've been better if the humanoid that was spawned wouldn't be as good as a round start species, otherwise you'd just be able to respawn once the botanist whips up a batch of them, at least with the current implementation there are downsides.
> 
> Either way with @metalgearsloth and others not wanting to add more stuff to botany until it is refactored makes me think its not worth the time to make changes to this PR unless there is a solution that makes everyone happy

yeah, that was the intention: however, diona also have a 'growth stage' post-harvest where they're a nymph and have to harvest blood samples in order to grow into a full humanoid - the growth stage is essentially the same as a monkey - doing something like this might be out of scope and would be a lot more to review for this PR, and i'm not sure if we need *more* species at the moment either

**Comment by ThunderBear2006** [comment]
Closing because botany needs a rework to both have better gameplay and to not be a code abomination

**Comment by Carmell0** [comment]
For the love of GOD add an industrial reagent grinder, litteraly just a blender with inf capacity and a chem master to put reagents in. This would be SO MUCH BETTER for botany

**Comment by Aisu9** [comment]
> For the love of GOD add an industrial reagent grinder, litteraly just a blender with inf capacity and a chem master to put reagents in. This would be SO MUCH BETTER for botany

just build a material reclaimer, it grinds stuff instantly and with no material inefficiency

**Comment by Qwazori** [comment]
Since the botany refactoring is already partially implemented and already well underway, I think it's worth resuming this pull request

**Pull request closed without merging**